### PR TITLE
[DataGrid] fix: use getRowId in filtering

### LINKED ISSUE #9562: Datagrid client-side filtering with custom id is broken in 6.9.1
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Steps to reproduce 🕹

Clientside filtering on @mui/x-data-grid 6.9.1 is broken when a custom id field is used.
```tsx
const columns: GridColDef[] = [
  {
    field: 'customId',
    headerName: 'Custom ID',
    type: 'number',
    width: 75,
  },
  {
    field: 'name',
    headerName: 'Name',
    type: 'string',
    width: 300,
  },
];

const rows = [
  {
    customId: 1,
    name: 'Hello',
  },
  {
    customId: 2,
    name: 'World',
  },
];

export default function App() {
  return (
    <DataGrid
      autoHeight
      columns={columns}
      getRowId={(row) => row.customId}
      rowCount={rows.length}
      rows={rows}
      slots={{
        toolbar: GridToolbar,
      }}
    />
  );
}

```
[Example on Stackblitz](https://stackblitz.com/edit/6-9-1-mui-datagrid-customid-column?file=src%2FApp.tsx)

### Current behavior 😯

Try to filter a column like the name on any value. The datagrid will not be filtered.

If the id column is named exactly "id" it is still working as expected:
[Example on Stackblitz](https://stackblitz.com/edit/6-9-1-mui-datagrid-id-column?file=src%2FApp.tsx)

### Expected behavior 🤔

Clientside filtering on column which are not named "id" should work.

### Context 🔦

_No response_

### Your environment 🌎

This issue is happening with @mui/x-data-grid 6.9.1, but not with 6.9.0

### Order ID or Support key 💳 (optional)

_No response_

### LINKED ISSUE #9586: Quick filter breaking on grouped rows & some rows with random ids in 6.9.1
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Steps to reproduce 🕹

Link to live example: https://codesandbox.io/s/competent-wiles-yvykm5?file=/demo.tsx

Steps:
1. Use the quick filter box to search any value
2. Observe error: 

Screenshot 2023-07-05 at 11 22 39 AM

Up to version 6.9.0, this behavior wasn't occurring.  The only indication I can see of what might have caused it is possibly the optimizations to filtering https://github.com/mui/mui-x/pull/9254, but I'm not sure.

### Current behavior 😯

Current behavior *appears* to be that when applying a quick (text) filter onto the table, rows that have derived ids (specified by a `getRowId` function) seem to cause issues, resulting in a  'no row found' error.  The two cases where I noticed it happening were:

-  when searching on rows where the id was a UUID generated by a `getRowId` function,
- when searching on rows in a table with some grouping applied, and a composite ID in `getRowId` (e.g [row.userId, row.cid].join('/')

### Expected behavior 🤔

I would expect that quick filtering should still (and always) work with any rows that have valid ids generated either by default (from the row.id field) or by the getRowId function.

### Context 🔦

This popped up as a bug ticket after we updated from 6.9.0 to 6.9.1.

### Your environment 🌎

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  System:
    OS: macOS 13.3.1
  Binaries:
    Node: 18.16.0 - ~/.asdf/installs/nodejs/18.16.0/bin/node
    Yarn: 3.5.1 - ~/.asdf/installs/yarn/1.22.19/bin/yarn
    npm: 9.5.1 - ~/.asdf/installs/nodejs/18.16.0/bin/npm
  Browsers:
    Chrome: 114.0.5735.198
    Edge: Not Found
    Firefox: 114.0
    Safari: 16.4
  npmPackages:
    @emotion/react: ^11.11.1 => 11.11.1 
    @emotion/styled: ^11.11.0 => 11.11.0 
    @mui/base:  5.0.0-beta.4 
    @mui/core-downloads-tracker:  5.13.4 
    @mui/icons-material: ^5.11.16 => 5.11.16 
    @mui/lab: ^5.0.0-alpha.134 => 5.0.0-alpha.134 
    @mui/material: ^5.13.6 => 5.13.6 
    @mui/private-theming:  5.13.1 
    @mui/styled-engine:  5.13.2 
    @mui/system:  5.13.5 
    @mui/types:  7.2.4 
    @mui/utils:  5.13.6 
    @mui/x-data-grid:  6.9.1 
    @mui/x-data-grid-premium: ^6.9.1 => 6.9.1 
    @mui/x-data-grid-pro:  6.9.1 
    @mui/x-date-pickers: ^6.9.1 => 6.9.1 
    @mui/x-license-pro: ^6.9.0 => 6.9.0 
    @types/react: ^18.2.7 => 18.2.7 
    react: ^18.2.0 => 18.2.0 
    react-dom: ^18.2.0 => 18.2.0 
    typescript: ^5.1.6 => 5.1.6 
```
</details>

Bug occurred while using Chrome

### Order ID or Support key 💳 (optional)

_No response_

### PR DESCRIPTION
**Pull request by romgrk**
Closes #9562
Closes #9586 

Use `getRowId` in filtering.

Issue with quick filter: https://codesandbox.io/s/competent-wiles-yvykm5?file=/demo.tsx

### CODE REVIEW

File: packages/grid/x-data-grid/src/hooks/features/filter/useGridFilter.tsx [hunk 1]
```diff
@@ -437,7 +438,7 @@ export const useGridFilter = (
         filteredDescendantCountLookup: {},
       };
     },
-    [apiRef, props.filterMode, rows],
+    [apiRef, props, rows],
```
Comment (MBilalShafi) [inline packages/grid/x-data-grid/src/hooks/features/filter/useGridFilter.tsx:441]: Wouldn't this save some unnecessary recomputations? (Keeping in mind that `getRowId` has a stable reference)
```suggestion
    [apiRef, props.filterMode, props.getRowId, rows],
```

Comment (romgrk) [reply packages/grid/x-data-grid/src/hooks/features/filter/useGridFilter.tsx]: The linter wanted `props` in the dependencies array for some reason. I've refactored to include those fields.

**Comment by cherniavskii** [comment]
Could you add a unit test to cover this regression issue?

**Comment by romgrk** [comment]
I've added the tests and a fix for another closely related issue.

**Review by cherniavskii — APPROVED** [review]

**Pull request merged**